### ISSUE TITLE
ACA validation fails due to exception caused when measured boot log size exceeds storage capacity.

### ISSUE THREAD

**Issue by Cyborg1138** [nsacyber/HIRS#489]
From the two ACA log extracts below it can be seen that provisioner posted measured boot logs can cause exceptions, resulting in validation failure. In such cases, no validation report is produced in the browser view, making it difficult to assess what happened.

The cause appears to be rimBytes being allocated SQL type blob as below. Excuse the formatting:)

This was found using a modified version of the .Net provisioner code base run on Windows 10. Blob is only good for 64K, but Windows measured boot logs can easily exceed this size. I have many that are >66K and as big as 77K on the machine in question.

MariaDB [hirs_db]> show columns from ReferenceManifest;
+-------------------------+--------------+------+-----+---------+-------+
| Field                   | Type         | Null | Key | Default | Extra |
+-------------------------+--------------+------+-----+---------+-------+
| DTYPE                   | varchar(31)  | NO   |     | NULL    |       |
| id                      | varchar(255) | NO   | PRI | NULL    |       |
| create_time             | datetime     | YES  |     | NULL    |       |
| archived_description    | varchar(255) | YES  |     | NULL    |       |
| archived_time           | datetime     | YES  |     | NULL    |       |
| associatedRim           | varchar(255) | YES  |     | NULL    |       |
| deviceName              | varchar(255) | YES  |     | NULL    |       |
| eventLogHash            | varchar(255) | YES  |     | NULL    |       |
| fileName                | varchar(255) | NO   |     | NULL    |       |
| hexDecHash              | varchar(255) | YES  |     | NULL    |       |
| platformManufacturer    | varchar(255) | YES  |     | NULL    |       |
| platformManufacturerId  | varchar(255) | YES  |     | NULL    |       |
| platformModel           | varchar(255) | YES  |     | NULL    |       |
| rimBytes                | blob         | NO   |     | NULL    |       |

**Extract 1:**

	at hirs.persist.DBManager.save(DBManager.java:184) [HIRS_Utils-2.1.1-SNAPSHOT.jar:?]
	at hirs.persist.DBReferenceManifestManager.save(DBReferenceManifestManager.java:23) [HIRS_Utils-2.1.1-SNAPSHOT.jar:?]
	at hirs.attestationca.AbstractAttestationCertificateAuthority.parseDeviceInfo(AbstractAttestationCertificateAuthority.java:942) [AbstractAttestationCertificateAuthority.class:?]
	at hirs.attestationca.AbstractAttestationCertificateAuthority.processDeviceInfo(AbstractAttestationCertificateAuthority.java:1067) [AbstractAttestationCertificateAuthority.class:?]
	at hirs.attestationca.AbstractAttestationCertificateAuthority.doSupplyChainValidation(AbstractAttestationCertificateAuthority.java:495) [AbstractAttestationCertificateAuthority.class:?]
	at hirs.attestationca.AbstractAttestationCertificateAuthority.processIdentityClaimTpm2(AbstractAttestationCertificateAuthority.java:435) [AbstractAttestationCertificateAuthority.class:?]
	at hirs.attestationca.rest.RestfulAttestationCertificateAuthority.processIdentityClaimTpm2(RestfulAttestationCertificateAuthority.java:98) [RestfulAttestationCertificateAuthority.class:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_342]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_342]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_342]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_342]
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:204) [spring-web-4.3.30.RELEASE.jar:4.3.30.RELEASE]
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:132) [spring-web-4.3.30.RELEASE.jar:4.3.30.RELEASE]
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:97) [spring-webmvc-4.3.30.RELEASE.jar:4.3.30.RELEASE]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:854) [spring-webmvc-4.3.30.RELEASE.jar:4.3.30.RELEASE]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:765) [spring-webmvc-4.3.30.RELEASE.jar:4.3.30.RELEASE]
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:85) [spring-webmvc-4.3.30.RELEASE.jar:4.3.30.RELEASE]
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:967) [spring-webmvc-4.3.30.RELEASE.jar:4.3.30.RELEASE]
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:901) [spring-webmvc-4.3.30.RELEASE.jar:4.3.30.RELEASE]
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970) [spring-webmvc-4.3.30.RELEASE.jar:4.3.30.RELEASE]
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:872) [spring-webmvc-4.3.30.RELEASE.jar:4.3.30.RELEASE]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:650) [tomcat-servlet-3.0-api.jar:?]
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846) [spring-webmvc-4.3.30.RELEASE.jar:4.3.30.RELEASE]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:731) [tomcat-servlet-3.0-api.jar:?]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303) [catalina.jar:7.0.76]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208) [catalina.jar:7.0.76]
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52) [tomcat7-websocket.jar:7.0.76]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241) [catalina.jar:7.0.76]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208) [catalina.jar:7.0.76]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:218) [catalina.jar:7.0.76]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:110) [catalina.jar:7.0.76]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:498) [catalina.jar:7.0.76]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:169) [catalina.jar:7.0.76]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:103) [catalina.jar:7.0.76]
	at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:962) [catalina.jar:7.0.76]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116) [catalina.jar:7.0.76]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:445) [catalina.jar:7.0.76]
	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1091) [tomcat-coyote.jar:7.0.76]
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:637) [tomcat-coyote.jar:7.0.76]
	at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:316) [tomcat-coyote.jar:7.0.76]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_342]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_342]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-coyote.jar:7.0.76]
	at java.lang.Thread.run(Thread.java:750) [?:1.8.0_342]
Caused by: java.sql.SQLDataException: (conn=2892) Data too long for column 'rimBytes' at row 1
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.get(ExceptionMapper.java:167) ~[mariadb-java-client-2.2.1.jar:?]
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.getException(ExceptionMapper.java:110) ~[mariadb-java-client-2.2.1.jar:?]
	at org.mariadb.jdbc.MariaDbStatement.executeExceptionEpilogue(MariaDbStatement.java:228) ~[mariadb-java-client-2.2.1.jar:?]
	at org.mariadb.jdbc.MariaDbPreparedStatementClient.executeInternal(MariaDbPreparedStatementClient.java:216) ~[mariadb-java-client-2.2.1.jar:?]
	at org.mariadb.jdbc.MariaDbPreparedStatementClient.execute(MariaDbPreparedStatementClient.java:150) ~[mariadb-java-client-2.2.1.jar:?]
	at org.mariadb.jdbc.MariaDbPreparedStatementClient.executeUpdate(MariaDbPreparedStatementClient.java:183) ~[mariadb-java-client-2.2.1.jar:?]
	at com.zaxxer.hikari.proxy.PreparedStatementProxy.executeUpdate(PreparedStatementProxy.java:61) ~[HikariCP-2.4.1.jar:?]
	at com.zaxxer.hikari.proxy.HikariPreparedStatementProxy.executeUpdate(HikariPreparedStatementProxy.java) ~[HikariCP-2.4.1.jar:?]
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:208) ~[hibernate-core-4.3.11.Final.jar:4.3.11.Final]
	... 60 more
Caused by: java.sql.SQLException: Data too long for column 'rimBytes' at row 1
Query is: insert into ReferenceManifest (create_time, archived_description, archived_time, associatedRim, deviceName, eventLogHash, fileName, hexDecHash, platformManufacturer, platformManufacturerId, platformModel, rimBytes, rimType, swidPatch, swidSupplemental, swidTagVersion, swidVersion, tagId, overallValidationResult, pcrHash, DTYPE, id) values (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 'EventLogMeasurements', ?), parameters ['2022-08-30 12:19:39.06','Event Log Measurement','2022-08-30 12:19:39.06',<null>,'host.docker.internal','','host.docker.internal.measurement','fdc5b69316d036be7a0a1ffa637d1df3616a67e212339845dccc684d643227c6','HP',<null>,'HP EliteBook 840 G5',<bytearray:\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00%\00\00\00Spec ID Event03\00\00\00\00\00\00\00\00\00\00\00\00\00 \00\00\00\00\00\00\00\00\00\00\00\00\00&��}�0~)-=� @9ܗ��\00�r�#T�'����_�20R��+I��w�e�R�\00\00Q78\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00��yNơtm����S"�mO�xӒnU�4���1v]
����:�
��P��$�!(�b�E˙�����d�J\00����-�ˑٳq��y^#�����Ɉ�^�\00Fg���%�v�k��q�>�N���E*u�!z����C%�Naa;�G�B���U��ƴ�@{�IN�9ҏL...

**Extract 2:**

2022-08-30 12:19:39 [hirs.attestationca.AbstractAttestationCertificateAuthority.processIdentityClaimTpm2] ERROR : java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
2022-08-30 12:19:39 [hirs.attestationca.AbstractAttestationCertificateAuthority.processIdentityClaimTpm2] ERROR : org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
2022-08-30 12:19:39 [hirs.attestationca.AbstractAttestationCertificateAuthority.processIdentityClaimTpm2] ERROR : java.lang.Thread.run(Thread.java:750)
2022-08-30 12:19:39 [hirs.attestationca.AbstractAttestationCertificateAuthority.processIdentityClaimTpm2] ERROR : Supply chain validation did not succeed. Result is: FAIL

**Comment by iadgovuser26** [comment]
What version of the ACA is being used? The storage size allocated for the RIM was updated in the [2.1.2 release](https://github.com/nsacyber/HIRS/releases) of the ACA. Please give that try. 

BTW the .Net provisioner is still in a development branch. Please keep this in mind as you test it out.

**Comment by Cyborg1138** [comment]
2.1.1. not the pre-release 2.1.2. 

We don't actually need that aspect to work at the moment and won't for a good while. I have prevented the provisioner from posting the measured boot log so it's not an issue. So we almost certainly won't try 2.1.2 for quite some time.

We are aware that the .Net provisioner is in a dev branch. I have modified the provisioner code quite a bit to get it working and it appears to do what we need it to do with these mods in. For use the advantages of using it are considerable, so that's the path we have chosen to take for the foreseeable future.

When the .Net provisioner gets merged into the master branch we will give it a try.
